src/EditorFeatures/Core/Implementation/IntelliSense/EditorCompletionImplementation/EditorAsyncCompletionItemManager.cs

```diff
@@ -0,0 +1,614 @@
+using Microsoft.VisualStudio.Language.Intellisense.AsyncCompletion;
+using Microsoft.VisualStudio.Text;
+using Roslyn.Utilities;
+using EditorCompletion = Microsoft.VisualStudio.Language.Intellisense.AsyncCompletion.Data;
+using RoslynCompletionItem = Microsoft.CodeAnalysis.Completion.CompletionItem;
+
+namespace Microsoft.CodeAnalysis.Editor.Implementation.IntelliSense.Completion.EditorImplementation
+{
+    internal class EditorAsyncCompletionItemManager : IAsyncCompletionItemManager
+    {
+        private readonly IAsyncCompletionBroker _broker;
+        private readonly CompletionHelper _completionHelper;
+
+        private const int MaxMRUSize = 10;
+        private ImmutableArray<string> _recentItems = ImmutableArray<string>.Empty;
+
+        public EditorAsyncCompletionItemManager(IAsyncCompletionBroker broker)
+        {
+            _broker = broker;
+            _completionHelper = new CompletionHelper(isCaseSensitive: true);
+        }
+
+        public Task<ImmutableArray<EditorCompletion.CompletionItem>> SortCompletionListAsync(
+            IAsyncCompletionSession session,
+            EditorCompletion.AsyncCompletionSessionInitialDataSnapshot data,
+            CancellationToken cancellationToken)
+        {
+            session.ItemCommitted += ItemCommitted;
+            session.Dismissed += SessionDismissed;
+            return Task.FromResult(data.InitialList.OrderBy(i => i.SortText).ToImmutableArray());
+        }
+
+        private void ItemCommitted(object sender, EditorCompletion.CompletionItemEventArgs e)
+        {
+            MakeMostRecentItem(e.Item.DisplayText);
+        }
+
+        private void SessionDismissed(object sender, EventArgs e)
+        {
+            if (sender is IAsyncCompletionSession session)
+            {
+                session.ItemCommitted -= ItemCommitted;
+                session.Dismissed -= SessionDismissed;
+            }
+        }
+
+        public Task<EditorCompletion.FilteredCompletionModel> UpdateCompletionListAsync(
+            IAsyncCompletionSession session,
+            EditorCompletion.AsyncCompletionSessionDataSnapshot data,
+            CancellationToken cancellationToken)
+        {
+            return Task.FromResult(UpdateCompletionList(session, data, cancellationToken));
+        }
+
+        public EditorCompletion.FilteredCompletionModel UpdateCompletionList(
+            IAsyncCompletionSession session,
+            EditorCompletion.AsyncCompletionSessionDataSnapshot data,
+            CancellationToken cancellationToken)
+        {
+            if (!session.Properties.TryGetProperty<bool>(CompletionItemSource.MustSetSelection, out bool mustSetSelection))
+            {
+                // This is the scenario when the session is created out of Roslyn, in some other provider, e.g. in Debugger.
+                // For now, the default mustSetSelection is false. Can discuss if the opposite is required.
+                mustSetSelection = false;
```

Comment:
i don't know what this means, because i don't know what "mustSetSelection" even does or controls...

Reply:
renamed to hasSuggestedItemOptions. This is initialized when we're calculating GetCompletionContextAsync

---
In reply to: [222085242](https://github.com/dotnet/roslyn/pull/29016#discussion_r222085242) [](ancestors = 222085242)